### ISSUE TITLE
Make available via Carthage

### ISSUE THREAD

**Issue by ianbytchek** [DevMate/DevMateKit#10]

**Comment by dimaty** [comment]
Thank you for your inquiry.
We consider making DevMateKit available via Carthage, but now we cannot provide any specific estimates when it will be done due to some technical difficulties. Please stay tuned to DevMateKit release notes.

**Comment by ianbytchek** [comment]
Hey @dimaty, that's good news. Is that because the source code is not available or something else??

**Comment by dimaty** [comment]
Not actually, that’s more about the specifics of DevMateKit integration process.

**Comment by ianbytchek** [comment]
@dimaty I just tried giving CocoaPods another go and now hate it even more. Hence double interested in getting this solved. 

I came across this [issue](http://stackoverflow.com/q/31901004/458356) today. After reading about it and nested frameworks I see that this is a weak point of Carthage, but CocoaPods comes out dry from the water. Is this what you are referring to by the integration process?

I know that some libraries are distributed as binaries in Carthage without need for compiling them, so I guess you can just use the same binary in downloads section. Using recursive option on the frameworks path solves the issue with nested DevMateIssuesReporter framework.

That's just me guessing, but if you can share some details perhaps I or someone else might suggest a working solution that will work for know. CocoaPods isn't in everyone's favourite list, but manually adding dependencies goes way beyond that. :confused:

**Comment by dimaty** [comment]
There is a problem with XPC service that should be added to copy phase of main target.

**Comment by ianbytchek** [comment]
Okay, but we don't have to do that when manually adding a binary? Why would that be a problem?

**Comment by dimaty** [comment]
When integrating DevMateKit manually and setting up updates feature there is a step where you have to create a special copy phase for XPC service (https://github.com/DevMate/DevMateKit/wiki/Updates). Using some kind of package managers (e.g. CocoaPods or Carthage), all necessary components are expected to be integrated by that manager.

**Comment by ianbytchek** [comment]
I see. I've been dealing with the MAS version only, so didn't get to that stage. So, did I get it right that Carthage not being able to add a build phase to Xcode project to install XPC components is the reason why this package is not provided via Carthage? Just wanna make sure I understand it :blush:

**Comment by dimaty** [comment]
Starting from v1.6.2 it will be available only for non-MAS version. Just modify your Cartfile by adding
`github "DevMate/DevMateKit" ~> 1.6.2`
And don't forget to go to the _'Build Settings'_ tab and add the following string to the _'Framework Search Paths'_:
`"$(PROJECT_DIR)/Carthage/Build/Mac/DevMateKit.framework/Frameworks"`
If you have more questions or suggestions we'll try to help you with it.

**Comment by ianbytchek** [comment]
Awesome. I assume non-MAS would work fine in a MAS as well, given I won't use any of it's features?

**Comment by dimaty** [comment]
Actually no. Non-MAS version contains bundles `com.devmate.UpdateInstaller.xpc`, `Autoupdate.app` and `Problem Reporter.app` that are not sandboxed. Because of that your apps will be rejected while reviewing. Another reason for rejection can be presence of activation/licensing and updating (Sparkle) API in framework. All these are absent in MAS version of DevMateKit. 

To make it available with Carthage for MAS version we could create additional directories for both frameworks (e.g. _'$(PROJECT_DIR)/Carthage/Build/Mac/DevMateKit/DevMateKit.framework'_ and _'$(PROJECT_DIR)/Carthage/Build/Mac/DevMateKit_MAS/DevMateKit.framework'_ that will be created after unarchiving _DevMateKit.framework.zip_) but I'm not sure that it's a correct solution according to Carthage documentation.

**Comment by ianbytchek** [comment]
Yes, now I recall reading about this in the docs. Can't say anything about the extra dirs, you could simply put them on the same level and call them `DevMateKit.framework` and `DevMateKit-MAS.framework`. [Documentation](https://github.com/Carthage/Carthage#archive-prebuilt-frameworks-into-one-zip-file) doesn't define any rules:

> To offer prebuilt frameworks for a specific tag, the binaries for all supported platforms should be zipped up together into one archive, and that archive should be attached to a published Release corresponding to that tag. The attachment should include .framework in its name (e.g., ReactiveCocoa.framework.zip), to indicate to Carthage that it contains binaries.

After working with Carthage for awhile and never saw any folders inside `Build/{Platform}`, I think it simply is clearer and easier to use frameworks when they are flat out.

Another point is that `update_reporter_id.sh` would also need to be included, since the framework would be useless without it. My suggestion is to put it on the top level `Script` folder and name it more specifically, like `dmk_reporter_id_updater.sh`. So the overall structure would be:

```
Build
    Mac
        DevMateKit.framework
        DevMateKit-MAS.framework
Script
    dmk_reporter_id_updater.sh
```

Nice and clean. What do you think?

**Comment by lukaskubanek** [comment]
@dimaty What is the state of this issue?
